docs/use/transform.md

```diff
@@ -0,0 +1,139 @@
+# Transforming specifications
+
+The `swagger` toolkit allows some transformation to be carried on with a specification.
+
+Currently it is possible to apply the following transforms:
+- expansion: this expands all `$ref`'s in a spec
+- minimal flattening: carries on minimal transformation to a spec to be workable for the swagger codegen
+- full flattening: performs minimal flattening and in addition, replaces all complex constructs in schemas by named definitions
+- mixin: merges one or more specifications into a primary spec
+
+### Expansion
+
+Expanding a spec may prove useful to produce documentation or test cases. The primary intent is not code generation.
+
+> **NOTE**: Circular `$ref` are detected and remain as local `$ref`
+
+Usage:
+
+`swagger expand {spec}`
+or
+
+`swagger flatten --with-expand` {spec}`
+
+or with codegen commands:
+
+`swagger generate [model|server|client|operation|...] --spec={spec} --with-expand`
+
+> **NOTE**: codegen may fail in some situations with spec expansion:
+> - polymorphism: the original intent of the `$ref` to a discriminated type(i.e. _extends type..._) is lost with expansion
+> - duplicate names: expansion of `$ref` intended for type reuse may produce duplicate type identifiers
+> - too many generated files: expansion of large specs may lead to many independent anonymous structures
+
+### Minimal flattening
+
+This transforms makes complex JSON `$ref` amenable to analysis and code generation.
+
+Minimal flattening attempts to minimally distorts the original specification intent, while ensuring workable codegen.
+
+Minimal flattening does:
+
+- bundle all external `$ref`s into the local document
+- resolve JSON pointers to anonymous places in the document (e.g. `"$ref": "#/definitions/thisModel/properties/codeName"`)
+- resolve `$ref`s in swagger-specific sections: `parameters` and `responses`, so the only remaining `$ref`s are located in schemas
+
+All `$ref`s in the resulting spec are thus only found in schema, with the _canonical_ form: `"$ref": "#/definitions/modelName"`.
+
+Usage:
+
+`swagger flatten {spec}`
+
+or more explicitly:
+
+`swagger flatten --with-flatten:minimal {spec}`
+
+This is the default option for codegen commands:
+
+`swagger generate [model|server|client|operation|...] --spec={spec}`
+
+> **NOTE**: `$ref` bundling / pointer resolving may in some cases produce duplicate names.
+> The flattener tries to resolve duplicates whenever possible (such as when a child identifier duplicates its parent).
+> When such resolution is not possible, an "OAIGen" suffix is added to the definition name (a warning is issued).
+> You may use the `x-go-name` extension here to set a more suitable name to the generated type.
```

Comment:
"a more suitable name for the generated type"?

Reply:
done